File: front/modules/youth-paper-widget/android/src/main/java/expo/modules/youthpaperwidget/YouthPaperWidgetModule.kt
package expo.modules.youthpaperwidget

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class YouthPaperWidgetModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("YouthPaperWidget")

    AsyncFunction("writeMealPayload") { json: String ->
      val ctx = requireContext()
      WidgetStore.write(ctx, WidgetStore.MEAL_KEY, json)
      WidgetUpdater.reloadAll(ctx)
    }

    AsyncFunction("writeTimetablePayload") { json: String ->
      val ctx = requireContext()
      WidgetStore.write(ctx, WidgetStore.TIMETABLE_KEY, json)
      WidgetUpdater.reloadAll(ctx)
    }

    AsyncFunction("writePeriodTimeSettings") { json: String ->
      val ctx = requireContext()
      WidgetStore.write(ctx, WidgetStore.PERIOD_KEY, json)
      WidgetUpdater.reloadAll(ctx)
    }

    AsyncFunction("writeSchoolId") { schoolId: String ->
      WidgetStore.write(requireContext(), WidgetStore.SCHOOL_ID_KEY, schoolId)
    }

    AsyncFunction("writeApiBaseUrl") { url: String ->
      WidgetStore.write(requireContext(), WidgetStore.API_BASE_URL_KEY, url)
    }

    AsyncFunction("writeAuthMirror") { json: String ->
      WidgetStore.write(requireContext(), WidgetStore.AUTH_MIRROR_KEY, json)
    }

    AsyncFunction("clearAuthMirror") {
      WidgetStore.remove(requireContext(), WidgetStore.AUTH_MIRROR_KEY)
    }

    AsyncFunction("readAuthMirror") {
      WidgetStore.read(requireContext(), WidgetStore.AUTH_MIRROR_KEY)
    }

    AsyncFunction("reloadWidgets") {
      WidgetUpdater.reloadAll(requireContext())
    }

    AsyncFunction("scheduleBackgroundRefresh") {
      val ctx = requireContext()
      WidgetRefreshScheduler.schedule(ctx)
      WidgetUpdater.reloadAll(ctx)
    }
  }

  private fun requireContext() =
    appContext.reactContext ?: throw IllegalStateException("React context lost")
}

File: front/modules/youth-paper-widget/ios/YouthPaperWidgetModule.swift
import ExpoModulesCore
import Foundation
import WidgetKit
import BackgroundTasks

public class YouthPaperWidgetModule: Module {
  public func definition() -> ModuleDefinition {
    Name("YouthPaperWidget")

    // AppDelegate 주입 대신 모듈 로드 시 등록 (Archive/EAS 컴파일 안전)
    OnCreate {
      WidgetBackgroundScheduler.register()
      WidgetBackgroundScheduler.scheduleAll()
    }

    AsyncFunction("writeMealPayload") { (json: String) in
      WidgetSharedStore.write(key: WidgetSharedStore.mealKey, value: json)
    }

    AsyncFunction("writeTimetablePayload") { (json: String) in
      WidgetSharedStore.write(key: WidgetSharedStore.timetableKey, value: json)
    }

    AsyncFunction("writePeriodTimeSettings") { (json: String) in
      WidgetSharedStore.write(key: WidgetSharedStore.periodTimeSettingsKey, value: json)
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadTimelines(ofKind: "TimetableWidget")
      }
    }

    AsyncFunction("writeSchoolId") { (schoolId: String) in
      WidgetSharedStore.write(key: WidgetSharedStore.schoolIdKey, value: schoolId)
    }

    AsyncFunction("writeApiBaseUrl") { (url: String) in
      WidgetSharedStore.write(key: WidgetSharedStore.apiBaseUrlKey, value: url)
    }

    AsyncFunction("writeAuthMirror") { (json: String) in
      WidgetAuthKeychain.write(json: json)
    }

    AsyncFunction("clearAuthMirror") {
      WidgetAuthKeychain.clear()
    }

    AsyncFunction("readAuthMirror") { () -> String? in
      WidgetAuthKeychain.read()
    }

    AsyncFunction("reloadWidgets") {
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }

    AsyncFunction("scheduleBackgroundRefresh") {
      WidgetBackgroundScheduler.scheduleAll()
    }
  }
}
